gateway/hostname.go

```diff
@@ -0,0 +1,623 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	cid "github.com/ipfs/go-cid"
+	namesys "github.com/ipfs/go-namesys"
+	"github.com/libp2p/go-libp2p/core/peer"
+	dns "github.com/miekg/dns"
+
+	mbase "github.com/multiformats/go-multibase"
+
+	nsopts "github.com/ipfs/interface-go-ipfs-core/options/namesys"
+)
+
+type HostnameConfig struct {
+	// NameSystem is the name system used to resolve TXT records for DNSLink hosts.
+	NameSystem namesys.NameSystem
```

Comment:
```suggestion
	// NameSystem is the name system used to resolve content paths starting with /ipns/id.
	// It finds  TXT records for DNSLink hosts and IPNS records for CID with libp2p-key in subdomains.
	NameSystem namesys.NameSystem
```

Reply:
@hacdias Thoughts on removing `NameSystem` and using `ResolvePath` instead?

This way it is up to `ResolvePath` implementation to decide if they support IPNS, DNSLink, and how they resolve them.

Removes a lot of API surface, makes it more flexible.

Reply:
@lidel I will try doing so.